### DISCUSSION TITLE
onCleanup of children componentes is called before onCleanup of parent

### DISCUSSION THREAD

**Discussion by gabrielmfern** [solidjs/solid#1757] [Q&A] [answered]
Lets say we had a component

```jsx
const Parent = (props) => {
  const [isCleaningUp, setIsCleaningUp] = createSignal(false);
  onMount(() => {
    setIsCleaningUp(false);
  });
  onCleanup(() => {
    setIsCleaningUp(true);
    console.log('only now cleaning up the parent');
  });

  return (
    <ParentContext.Provider value={[isCleaningUp]}>
      {props.children}
    </ParentContext.Provider>
  );
};
```

And lets say that there were some children components inside of it:

```jsx
const Example = (props) => {
  const [isParentCleaningUp] = useContext(ParentContext);

  onMount(() => {
    console.log(isParentCleaningUp());
  });

  onCleanup(() => {
    if (isParentCleaningUp()) {
      console.log('should appear in the console');
      // ... do something differently than normal, maybe saving some 
      // kind of data that the component is responsible of managing
    } else {
      console.log('should not appear in the console');
      // other behaviour 
    }
  });

  return <div>{isParentCleaningUp().toString()}</div>;
}
```

What happens is that the children are cleaned up before the parent is shown when rendering the following:

```jsx
function App() {
  const [shouldShowParent, setShouldShowParent] = createSignal(true);
  onMount(() => {
    setTimeout(() => setShouldShowParent(false), 2000);
  });
  return <Show when={shouldShowParent()}>
    <Parent >
      <Example />
      <Example />
      <Example />
    </Parent>
  </Show>;
}
```

The above outputs the following in the console:
```
(3) false
(3) should not appear in the console
only now cleaning up the parent
```

And I think it makes sense that the onCleanup of the children is called first since Solid may clean up the component tree from bottom to top, but I still need to know that the context is being cleaned up and not only the Example. 

This can be done through the App, but it is really not optimal. How can this be done inside of the Parent?

**Comment by snnsnn** [comment]
You expect resources cleaned in reverse acquisition order. So, resources obtained by the children should be released first than you can safely release the parent's resources. If you look carefully components are themselves resources created by their parents.

There are several issues and discussion around this topic. For example:

https://github.com/solidjs/solid/issues/1562

**Reply by gabrielmfern** [reply]
I understand, but I don't really want the order to be different.
I just wanted to run something before the clean up of the `<Parent>` so that the children can have some
kind of differing behavior if the parent is being cleaned up. This would work regardless of the order of clean up.

**Reply by snnsnn** [reply]
You can use createRoot for the children and dispose them manually anytime you like.

**Reply by gabrielmfern** [reply]
>You can use createRoot for the children and dispose them manually anytime you like.

But still I need to know that, at the clean up of the children, they are being cleaned up due to the parent needing to be cleaned up. This doesn't really solve it, and the only way I could clean up the children with `createRoot` would be probably after the clean up of the Parent which would be unsafe.

**Reply by snnsnn** [reply, marked as answer]
```tsx
import {
  Show,
  createContext,
  createRoot,
  createSignal,
  onCleanup,
  onMount,
  useContext
} from 'solid-js';
import { render } from 'solid-js/web';

const ParentContext = createContext<() => boolean>(() => false);

const Parent = (props) => {
  let dispose: () => void;
  const [isCleaningUp, setIsCleaningUp] = createSignal(false);

  onMount(() => {
    setIsCleaningUp(false);
  });

  onCleanup(() => {
    setIsCleaningUp(true);
    console.log('Cleaning up parent');
    dispose && dispose();
  });

  const child = createRoot((disposer) => {
    dispose = disposer;
    return props.children;
  });

  return (
    <ParentContext.Provider value={isCleaningUp}>
      {child}
    </ParentContext.Provider>
  );
};

const Example = (props) => {
  const isParentCleaningUp = useContext(ParentContext);

  onMount(() => {
    console.log(isParentCleaningUp());
  });

  onCleanup(() => {
    console.log('Cleaning up children');
  });

  return <div>{isParentCleaningUp().toString()}</div>;
};

function App() {
  const [shouldShowParent, setShouldShowParent] = createSignal(true);
  onMount(() => {
    setTimeout(() => setShouldShowParent(false), 2000);
  });
  return (
    <Show when={shouldShowParent()}>
      <Parent>
        <Example />
        <Example />
        <Example />
      </Parent>
    </Show>
  );
}

render(App, document.body);
```

You are asking for children being cleaned up after their parent and that is how you achieve it. Since you keep the root inside parent component, you will have a brief time window inside onCleanup to do some work when the parent is unmounted,  which will not be enough to juggle states, let alone update UI.

If you want to keep the values or children even after parent unmounts, you need to elevate the state to an outer parent.

Now being under a different root, childrens are not cleaned up until we call the dispose callback. Comment out the line we dispose the root, you will observer `Cleaning up children` will never be logged.

If you want to persist the user input, there are many ways to do it efficiently. Few options:

1. You can create an external state for the input values, any state that lives outside parent, because outer Show will unmount the parent, whatever you have under it will be gone.

2. You can persist whatever user enters into localstorage and keep there until form is submitted. If somehow page is refreshed, you can restore values from the local storage. You can clean the values from local storage once the form is submitted.

**Reply by gabrielmfern** [reply]
Thank you very much for the answer.

But there is still a small problem that I am not sure of why it would happen.
For a reason that I am not aware of, the `isParentCleaningUp` signal remains `false` when the onCleanup of the children runs. Any idea of why this happens?

**Reply by snnsnn** [reply]
> But there is still a small problem that I am not sure of why it would happen. For a reason that I am not aware of, the `isParentCleaningUp` signal remains `false` when the onCleanup of the children runs. Any idea of why this happens?

Because Parent is removed and it never gets a chance to update the state. In other words when Show toggles off there wont be <ParentContext.Provider value={isCleaningUp}>.

> You will have a brief time window inside onCleanup to do some work when the parent is unmounted, which will not be enough to juggle states, let alone update UI.

**Reply by gabrielmfern** [reply]
Would that mean that if the context's provider value did not have default it would be undefined at the onCleanup of the children?

**Reply by gabrielmfern** [reply]
Okay, playing around a bit in the playground your solution works end-to-end, thank you. 
The problem was that the createRoot was not inside of the context and so the `useContext` just tried using the default, so even the state is being updated (as far as I tested).
